File: JavaScript/src/emailSender.js
export function EmailSender() {
    this.emails = [];
    this.observers = [];
}

EmailSender.prototype.register = function(observer) {
    this.observers.push(observer);
}
EmailSender.prototype.unregister = function(observer) {
    this.observers = this.observers.filter(ob => ob != observer);
}
EmailSender.prototype.notifyUpdates = function() {
    this.observers.forEach(observer => {
        observer.update(this);
    });
}
EmailSender.prototype.pushEmail = function(email) {
    console.log("%c***********************", 'font-weight:bolder;')
    this.emails.push(email);
    this.notifyUpdates();
    this.emails = [];
}

File: Java/src/main/java/vehicle/EmailSender.java
package vehicle;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Soufyan & Omar
 */
public class EmailSender implements Subject{
    private List<Email> emails;
    private List<Observer> observers;

    public EmailSender() {
        this.emails = new ArrayList<>();
        this.observers = new ArrayList<>();
    }

    @Override
    public void pushEmail(Email email){
        System.out.println("***********************");
        this.emails.add(email);
        this.notifyUpdates();
    }


    public List<Observer> getObservers() {
        return observers;
    }

    @Override
    public void register(Observer ob) {
        observers.add(ob);
    }

    @Override
    public void unregister(Observer ob) {
        observers.remove(ob);
    }

    @Override
    public void notifyUpdates() {
        for (Observer observer: observers) {
            observer.update(this);
        }
    }

    @Override
    public List<Email> getEmails() {
        return emails;
    }
}
